src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,316 @@
+            var patternMatcher = PatternMatcher.CreatePatternMatcher(pattern, includeMatchedSpans: false, allowFuzzyMatching: true);
+
+            foreach (var documentSymbol in documentSymbolItems)
+            {
+                if (SearchNodeTree(documentSymbol, patternMatcher, cancellationToken))
+                    documentSymbols.Add(documentSymbol);
+            }
+
+            return documentSymbols.ToImmutableAndFree();
+
+            // Returns true if the name of one of the tree nodes results in a pattern match.
+            static bool SearchNodeTree(DocumentSymbolItem tree, PatternMatcher patternMatcher, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (patternMatcher.Matches(tree.Name))
+                    return true;
+
+                foreach (var childItem in tree.Children)
+                {
+                    if (SearchNodeTree(childItem, patternMatcher, cancellationToken))
+                        return true;
+                }
+
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Returns the Document Symbol node that is currently selected by the caret in the editor if it exists.
+        /// </summary>
+        public static DocumentSymbolItem? GetDocumentNodeToSelect(
+            ImmutableArray<DocumentSymbolItem> documentSymbolItems,
+            ITextSnapshot originalSnapshot,
+            ITextSnapshot currentSnapshot,
+            int caretPosition)
+        {
+            return GetNodeToSelect(documentSymbolItems, null);
+
+            DocumentSymbolItem? GetNodeToSelect(ImmutableArray<DocumentSymbolItem> documentSymbols, DocumentSymbolItem? parent)
+            {
+                var selectedSymbol = GetNodeSelectedByCaret(documentSymbols);
+
+                if (selectedSymbol is null)
+                    return parent;
+
+                return GetNodeToSelect(selectedSymbol.Children, selectedSymbol);
+            }
+
+            // Returns a DocumentSymbolItem if the current caret position is in its range and null otherwise.
+            DocumentSymbolItem? GetNodeSelectedByCaret(ImmutableArray<DocumentSymbolItem> documentSymbolItems)
+            {
+                foreach (var symbol in documentSymbolItems)
+                {
+                    var originalStartPosition = originalSnapshot.GetLineFromLineNumber(symbol.StartPosition.Line).Start.Position + symbol.StartPosition.Character;
+                    var originalEndPosition = originalSnapshot.GetLineFromLineNumber(symbol.EndPosition.Line).Start.Position + symbol.EndPosition.Character;
+
+                    var originalSpan = new SnapshotSpan(originalSnapshot, Span.FromBounds(originalStartPosition, originalEndPosition));
+                    var currentSpan = originalSpan.TranslateTo(currentSnapshot, SpanTrackingMode.EdgeExclusive);
```

Comment:
this is doing all this work *per* item in the tree.  instead of this, as the very first thing, just map the current caret position *back* to the original in the *original* snapshot.  Then you can compare it to the span taht is in the data model.

also, looking at this code (and the prior code that chcks against line/columns), we really should just preemptive translate hte line/column passed back by LSP into a standard doc position *immediately* when we get the lsp request back.  While LSP is line/column based, roslyn and the editor are not (it's always positional).  SO it will be cleaner to map to that domain directly and then use that representation for all further work.